File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys

N = int(sys.stdin.readline())
ans = 0
row = [-1] * N

def is_safe(n):
    for i in range(n):
        if row[n] == row[i] or abs(row[n] - row[i]) == abs(n - i):
            return False

    return True


def n_queen(n):
    global ans
    if n == N:
        ans += 1
        return

    else:
        for i in range(N):
            row[n] = i
            if is_safe(n):
                n_queen(n + 1)

n_queen(0)
print(ans)

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.util.Scanner;

public class Main {
    private static int N;
    private static int[] row;
    private static int count;

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        row = new int[N];

        nQueen(0);
        System.out.println(count);
    }

    private static void nQueen(int n) {
        if (n == N){
            count++;
            return;
        }

        for (int i = 0; i < N; i++){
            row[n] = i;
            if (isSafe(n)) {
                nQueen(n + 1);
            }
        }
    }

    private static boolean isSafe(int n) {
        for (int i = 0; i < n; i++){
            if (row[n] == row[i] || Math.abs(row[n] - row[i]) == Math.abs(n - i)){
                return false;
            }
        }
        return true;
    }
}
